pkg/visibility/api/pending_workload_CQ.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2023 The Kubernetes Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package api
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	metainternalversion "k8s.io/apimachinery/pkg/apis/meta/internalversion"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	metav1beta1 "k8s.io/apimachinery/pkg/apis/meta/v1beta1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apiserver/pkg/registry/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	v1alpha1 "sigs.k8s.io/kueue/apis/api_extension/v1alpha1"
+	"sigs.k8s.io/kueue/pkg/queue"
+
+	_ "k8s.io/metrics/pkg/apis/metrics/install"
+)
+
+type pendingWorkloadsInCQ struct {
+	kueueMgr *queue.Manager
+	log      logr.Logger
+}
+
+const (
+	WholeSnapshot = 100000
+)
+
+var _ rest.KindProvider = &pendingWorkloadsInCQ{}
+var _ rest.Storage = &pendingWorkloadsInCQ{}
+var _ rest.Getter = &pendingWorkloadsInCQ{}
+var _ rest.Lister = &pendingWorkloadsInCQ{}
+var _ rest.Scoper = &pendingWorkloadsInCQ{}
+var _ rest.TableConvertor = &pendingWorkloadsInCQ{}
+var _ rest.SingularNameProvider = &pendingWorkloadsInCQ{}
+
+func newPendingWorkloadsInCQ(kueueMgr *queue.Manager) *pendingWorkloadsInCQ {
+	return &pendingWorkloadsInCQ{
+		kueueMgr: kueueMgr,
+		log:      ctrl.Log.WithName("pending-workload-cq"),
+	}
+}
+
+// New implements rest.Storage interface
+func (m *pendingWorkloadsInCQ) New() runtime.Object {
+	return &v1alpha1.PendingWorkloadSummary{}
+}
+
+// Destroy implements rest.Storage interface
+func (m *pendingWorkloadsInCQ) Destroy() {
+}
+
+// Kind implements rest.KindProvider interface
+func (m *pendingWorkloadsInCQ) Kind() string {
+	return "PendingWorkloadsInClusterQueue"
+}
+
+// NewList implements rest.Lister interface
+func (m *pendingWorkloadsInCQ) NewList() runtime.Object {
+	return &v1alpha1.PendingWorkloadSummaryList{}
+}
+
+// List implements rest.Lister interface
+func (m *pendingWorkloadsInCQ) List(ctx context.Context, options *metainternalversion.ListOptions) (runtime.Object, error) {
+	// TODO: consider returning error
+	return &v1alpha1.PendingWorkloadSummaryList{}, nil
+}
+
+// Get implements rest.Getter interface
+func (m *pendingWorkloadsInCQ) Get(ctx context.Context, name string, opts *metav1.GetOptions) (runtime.Object, error) {
+	m.kueueMgr.UpdateSnapshot(name, WholeSnapshot)
```

Comment:
IIUC, updating the snapshot might be problematic, as this would cause the visibility in status to save this snapshot. 

I think we could expose a new function by manager "ComputeHead(cqName, count) []*workload.Info". Then, call it from the new code. It means that the snapshot isn't updated, but I think this is ok. I think then the new function could also be reused by the existing code.

Reply:
Good point, fixed